### PR TITLE
[XCVM-69] feat(cosmwasm): introduce the native gate feature - CosmWasm <-> Pallet interactions

### PR DESCRIPTION
**Pull request by hussein-aitlahcen**
Introduce a native gate hook in the cosmwasm pallet, allowing us to use "pallets as contract". This feature also open the door to pallet acting as IBC capable (pablo over IBC directly) without having to rely on an actual cosmwasm contract implementation.

### CODE REVIEW

File: code/parachain/frame/cosmwasm/src/hook.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
+use crate::{runtimes::vm::CosmwasmVM, types::*, Config, Error};
+use cosmwasm_vm::{
+	cosmwasm_std::{ContractResult, QueryResponse, Response},
+	vm::{VMBase, VmErrorOf},
+};
+use cosmwasm_vm_wasmi::WasmiVM;
+
+/// A hook into the VM, used to call precompiled contracts.
+pub trait Hook<T: Config> {
```
Comment (KaiserKarel) [inline code/parachain/frame/cosmwasm/src/hook.rs:9]: Nit: a hook doesn't necessarily need to forward to precompiles; we can split this into two traits:

PrecompiledInfo + Hook

and remove the precompiled_prefix from member functions.

**Review by cor — APPROVED** [review]
Pair programmed with @hussein-aitlahcen

**Review by dzmitry-lahoda — APPROVED** [review]
as i see will need add more variants to do IBC hooks

**Comment by cor** [comment]
Rebased to include the v0.9.30 polkadot upgrade. Weights have been changed to just `u64` for now. We will have to reconsider our approach to gasmetering when Weights V2 gets released.

Also refactored to remove two useless abstractions: `CallerState` and `EntryPointCaller`. They provided no shared behavior, and only introduced more type complexity

**Review by aeryz — APPROVED** [review]

File: code/parachain/frame/cosmwasm/src/entrypoint.rs [hunk 2]
```diff
@@ -1,306 +1,136 @@
 use crate::{
```
[the remark this answers is not in the document]
Comment (cor) [reply code/parachain/frame/cosmwasm/src/entrypoint.rs:1]: I agree, but let's do it in a future PR as this one has been in the PR queue for a long time now

**Pull request merged**